### PR TITLE
Feature/rasterio

### PR DESCRIPTION
**Pull request by NicWayand**
@jhamman started a backend for RasterIO that I have been working on. There are two issues I am stuck on that I could use some help:

1) Lat/long coords are not being decoded correctly (missing from output dataset). Lat/lon projection are correctly calculated and added here (https://github.com/NicWayand/xray/blob/feature/rasterio/xarray/backends/rasterio_.py#L117). But, it appears (with my limited knowledge of xarray) that the lat/long coords contained within `obj` are lost at this line (https://github.com/NicWayand/xray/blob/feature/rasterio/xarray/conventions.py#L930). 

2) Lazy-loading needs to be enabled. How can I setup/test this? Are there examples from other backends I could follow?

#790

### CODE REVIEW

File: xarray/backends/rasterio_.py [hunk 1]
```diff
@@ -0,0 +1,138 @@
...
+            raise ValueError('unknown dims')
+
+        self._attrs = OrderedDict()
+        for attr_name in ['crs', 'transform', 'proj']:
+            try:
+                self._attrs[attr_name] = getattr(self.ds, attr_name)
+            except AttributeError:
+                pass
+
+        self.coords = _try_to_get_latlon_coords(self.coords, self._attrs)
```
Comment (shoyer) [inline xarray/backends/rasterio_.py]: You need to return these in the dict returned by `get_variables()`. `DataStore` objects don't have a `coords` attribute.

File: ci/requirements-py27-cdat+pynio.yml [hunk 2]
```diff
@@ -3,7 +3,7 @@ channels:
   - conda-forge
 dependencies:
   - python=2.7
-  - cdat-lite
+  # - cdat-lite
```
Comment (shoyer) [inline ci/requirements-py27-cdat+pynio.yml]: This seems unrelated?

**Comment by fmaussion** [comment]
This is awesome! 

For 2: I think you don't have much more to do than make ``__getitem_`` use the indexer to read the data with a window (example [here](https://github.com/fmaussion/salem/blob/master/salem/datasets.py#L341-L353))

File: xarray/backends/rasterio_.py (hunk 1 above)
Comment (NicWayand) [inline xarray/backends/rasterio_.py]: Thanks. This now pulls in lat/long if the projection was successful, otherwise it just returns the basic x/y indices. Should we also be returning the original projection coords? (currently they are lost).

File: xarray/backends/rasterio_.py [hunk 3]
```diff
@@ -0,0 +1,139 @@
...
+    coords_out = {}
+    try:
+        import pyproj
+    except ImportError:
+        pyproj = False
+    if 'crs' in attrs and pyproj:
+        proj = pyproj.Proj(attrs['crs'])
+        x, y = np.meshgrid(coords['x'], coords['y'])
+        proj_out = pyproj.Proj("+init=EPSG:4326", preserve_units=True)
+        xc, yc = _transform_proj(proj, proj_out, x, y)
```
Comment (shoyer) [inline xarray/backends/rasterio_.py]: You might try to do this calculation lazily, e.g., by making a `utils.NDArrayMixin` subclass in conventions.py and wrapping with `indexing.LazilyIndexedArray`.

File: xarray/backends/rasterio_.py [hunk 4]
```diff
@@ -0,0 +1,139 @@
...
+            raise ValueError('unknown dims')
+
+        self._attrs = OrderedDict()
+        for attr_name in ['crs', 'transform', 'proj']:
+            try:
+                self._attrs[attr_name] = getattr(self.ds, attr_name)
+            except AttributeError:
+                pass
+
+    # def get_vardata(self, var_id=1):
```
Comment (shoyer) [inline xarray/backends/rasterio_.py]: should be removed

File: xarray/backends/rasterio_.py [hunk 5]
```diff
@@ -0,0 +1,139 @@
...
+        # Get dims
+        if self.ds.count >= 1:
+            self.dims = ('band', 'y', 'x')
+            self.coords['band'] = self.ds.indexes
+        else:
+            raise ValueError('unknown dims')
+
+        self._attrs = OrderedDict()
+        for attr_name in ['crs', 'transform', 'proj']:
+            try:
```
Comment (shoyer) [inline xarray/backends/rasterio_.py]: use `with pycompat.suppress(AttributeError):`

File: xarray/backends/rasterio_.py [hunk 6]
```diff
@@ -0,0 +1,139 @@
...
+        else:
+            return copy.deepcopy(x), copy.deepcopy(y)
+
+    return pyproj.transform(p1, p2, x, y)
+
+
+def _try_to_get_latlon_coords(coords, attrs):
+    coords_out = {}
+    try:
+        import pyproj
```
Comment (shoyer) [inline xarray/backends/rasterio_.py]: I'm a little surprised rasterio doesn't have projections built in.

File: ci/requirements-py35.yml [hunk 7]
```diff
@@ -11,6 +11,7 @@ dependencies:
   - pandas
   - seaborn
   - scipy
+  - rasterio
```
Comment (shoyer) [inline ci/requirements-py35.yml]: Unless rasterio only supports Python 3, should add it to other test suites, too (especially Python 2.7).

**Comment by NicWayand** [comment]
Tested open_mfdataset() on 100+ geotiffs and lazyloading with rasterio does appear to be working.

File: ci/requirements-py35.yml (hunk 7 above)
Comment (NicWayand) [inline ci/requirements-py35.yml]: Rasterio supports 2.7 and 3.3-3.5 (https://github.com/mapbox/rasterio)

I added this requirement to all other test suites, was that correct? Version 1.0 is needed as many function names changed in rasterio at V1.0.

**Comment by NicWayand** [comment]
Travis-ci fail is because it can't find rasterio, which comes through the conda-forge channel (https://github.com/conda-forge/rasterio-feedstock). I think it needs to be added as described here (http://conda.pydata.org/docs/travis.html#additional-steps). But I am new to Travis-ci, so don't want to mess up the current .travis.yml file.

**Comment by shoyer** [comment]
Add `conda-forge` it to the `channels` section of conda requirements yml files:
https://github.com/pydata/xarray/blob/f40d323c76f6cab9615567e4ba7afde7860f7775/ci/requirements-py27-cdat%2Bpynio.yml#L3

**Comment by NicWayand** [comment]
Any idea why Segmentation faults occur for 3.4 and 4.5?

File: xarray/backends/rasterio_.py (hunk 6 above)
Comment (NicWayand) [inline xarray/backends/rasterio_.py]: They do: https://mapbox.github.io/rasterio/topics/reproject.html

I will take a stab at doing it the rasterio way.

File: xarray/backends/rasterio_.py (hunk 3 above)
Comment (NicWayand) [inline xarray/backends/rasterio_.py]: Sorry I don't understand how to do this. Is there an example I can go off of?

**Comment by shoyer** [comment]
I would start by adding `-v` to our invocation of `py.test` in `.travis.yml`: https://github.com/pydata/xarray/blob/a247d9314d6a8df34df887106e234bbffebe1e48/.travis.yml#L72

File: xarray/backends/rasterio_.py (hunk 3 above)
Comment (shoyer) [inline xarray/backends/rasterio_.py]: We have some examples of making NDArrayMixin subclasses in conventions.py, e.g.,
https://github.com/pydata/xarray/blob/master/xarray/conventions.py#L311

Comment (shoyer) [inline xarray/backends/rasterio_.py]: Another option would be to use `dask.array` to build the coordinate arrays, though the `NDArrayMixin` would avoid the need for dask.

File: xarray/backends/rasterio_.py [hunk 8]
```diff
@@ -0,0 +1,120 @@
...
+        self.array = ds.read()
+
+    @property
+    def dtype(self):
+        return np.dtype(self._ds.dtypes[0])
+
+    def __getitem__(self, key):
+        if key == () and self.ndim == 0:
+            return self.array.get_value()
+        return self.array[key]
```
Comment (jhamman) [inline xarray/backends/rasterio_.py]: Based on what @fmaussion said above, I think this should be something like:

```Python
    def __getitem__(self, key):
        if key == () and self.ndim == 0:
            return self._ds.read()
        return self._ds.read(band, window=window)
```

Where `band` and `window` describe the slice of data to be read. Off the top of my head, I'm not exactly sure how to parse the key here though.

File: xarray/backends/rasterio_.py (hunk 6 above)
Comment (shoyer) [inline xarray/backends/rasterio_.py]: That's a reprojection, which does seem useful, but it seems separate from giving you `lat` and `lon` coordinates

File: xarray/backends/rasterio_.py (hunk 8 above)
Comment (shoyer) [inline xarray/backends/rasterio_.py]: By the time indexers get here, they should have already passed through [canonicalize_indexer](https://github.com/pydata/xarray/blob/a247d9314d6a8df34df887106e234bbffebe1e48/xarray/core/indexing.py#L41), which means you only need to handle a `key` that is a tuple of the appropriate number of dimensions (i.e., 3) composed of integers, slices and integer ndarrays.

Based on the [docstring for read](https://github.com/mapbox/rasterio/blob/a5376390829be52490a1fe6da52395262d0f6167/rasterio/_io.pyx#L234-L287), we want something like:

```python
def __getitem__(self, key):
    band = range(self.shape[0])[key[0]]
    window = []
    for win in key[1:]:
        if instance(win, slice):
            window.append((win.start, win.stop))
        elif isinstance(win, int):
            window.append((win, win + 1))
        else:  # integer ndarray
            window.append((win.min(), win.max()))
    raw_data = self._ds.read(band, window=window)
    # now, fix up raw_data to conform to numpy indexing conventions
    # - drop axes for integer band/windows
    # - stride if windows are slices with win.step != 1
    # - subset if window is an integer ndarray
```

Honestly, this logic should probably live in rasterio if possible. I'm a little surprised that they have never implemented a `__getitem__` method.

Comment (shoyer) [inline xarray/backends/rasterio_.py]: See also: https://gist.github.com/lpinner/bd57b54a5c6903e4a6a2 (can't reuse this directly, though, because it doesn't have a license).

Anyways, I would definitely see if the rasterio folks are up for implementing a `__getitem__` method. There's no reason why this is xarray specific, to be honest.

Comment (jhamman) [inline xarray/backends/rasterio_.py]: See also https://github.com/mapbox/rasterio/issues/920 for related discussion.

File: ci/requirements-py27-min.yml [hunk 9]
```diff
@@ -1,9 +1,12 @@
 name: test_env
+channels:
+  - conda-forge
 dependencies:
   - python=2.7
   - pytest
   - numpy==1.9.3
   - pandas==0.15.0
+  - rasterio
```
Comment (jhamman) [inline ci/requirements-py27-min.yml]: I think we can roll back these changes.  This test setup is just for required dependencies which `rasterio` will not be.

File: ci/requirements-py27-netcdf4-dev.yml [hunk 10]
```diff
@@ -8,6 +10,7 @@ dependencies:
   - numpy
   - pandas
   - scipy
+  - rasterio
```
Comment (jhamman) [inline ci/requirements-py27-netcdf4-dev.yml]: I think we can also roll back these changes as well

File: ci/requirements-py33.yml [hunk 11]
```diff
@@ -1,8 +1,11 @@
 name: test_env
+channels:
+  - conda-forge
 dependencies:
   - python=3.3
   - pytest
   - pandas
+  - rasterio
```
Comment (jhamman) [inline ci/requirements-py33.yml]: roll back these changes

File: ci/requirements-py27-pydap.yml [hunk 12]
```diff
@@ -8,6 +10,7 @@ dependencies:
   - numpy
   - pandas
   - scipy
+  - rasterio
```
Comment (jhamman) [inline ci/requirements-py27-pydap.yml]: I don't think we need to test rasterio along with pydap so we can roll back these changes.

File: ci/requirements-py35-dask-dev.yml [hunk 13]
```diff
@@ -7,6 +9,7 @@ dependencies:
   - pandas
   - scipy
   - toolz
+  - rasterio
```
Comment (jhamman) [inline ci/requirements-py35-dask-dev.yml]: roll back these changes

File: ci/requirements-py35-pandas-dev.yml [hunk 14]
```diff
@@ -7,6 +9,7 @@ dependencies:
   - netcdf4=1.1.9
   - scipy=0.16.0
   - toolz
+  - rasterio
```
Comment (jhamman) [inline ci/requirements-py35-pandas-dev.yml]: roll back these changes

File: xarray/backends/rasterio_.py [hunk 15]
```diff
@@ -0,0 +1,120 @@
...
+
+from .common import AbstractDataStore
+
+__rio_varname__ = 'raster'
+
+
+class RasterioArrayWrapper(NDArrayMixin):
+    def __init__(self, ds):
+        self._ds = ds
+        self.array = ds.read()
```
Comment (jhamman) [inline xarray/backends/rasterio_.py]: @shoyer, do we need to maintain the `array` attribute here? Would it make more sense to just populate the `_ds` attribute and set the array in `__getitem__`?

File: xarray/backends/rasterio_.py [hunk 16]
```diff
@@ -0,0 +1,120 @@
...
+        if self.ds.count >= 1:
+            self.dims = ('band', 'y', 'x')
+            self.coords['band'] = self.ds.indexes
+        else:
+            raise ValueError('unknown dims')
+
+        self._attrs = OrderedDict()
+        with suppress(AttributeError):
+            for attr_name in ['crs', 'transform', 'proj']:
+                    self._attrs[attr_name] = getattr(self.ds, attr_name)
```
Comment (jhamman) [inline xarray/backends/rasterio_.py]: I think there is some extra indentation here -- should be 4 spaces.

File: xarray/backends/rasterio_.py [hunk 17]
```diff
@@ -0,0 +1,120 @@
...
+        data = indexing.LazilyIndexedArray(
+            RasterioArrayWrapper(self.ds))
+        return Variable(self.dims, data, self._attrs)
+
+    def get_variables(self):
+        # Get lat lon coordinates
+        coords = _try_to_get_latlon_coords(self.coords, self._attrs)
+        vars = {__rio_varname__: self.open_store_variable(__rio_varname__)}
+        vars.update(coords)
+        return FrozenOrderedDict(vars)
```
Comment (jhamman) [inline xarray/backends/rasterio_.py]: we should use another name other than `vars` since it is a python builtin.

File: xarray/test/test_backends.py [hunk 18]
```diff
@@ -1063,6 +1063,35 @@ def test_weakrefs(self):
...
+    def test_weakrefs(self):
+        example = Dataset({'foo': ('x', np.arange(5.0))})
+        expected = example.rename({'foo': 'bar', 'x': 'y'})
+
+        with create_tmp_file() as tmp_file:
+            example.to_netcdf(tmp_file, engine='scipy')
+            on_disk = open_dataset(tmp_file, engine='rasterio')
+            actual = on_disk.rename({'foo': 'bar', 'x': 'y'})
+            del on_disk  # trigger garbage collection
+            self.assertDatasetIdentical(actual, expected)
```
Comment (jhamman) [inline xarray/test/test_backends.py]: @NicWayand - take a look at the PyNio tests.  They are probably the closest analog for what we need to test here.

File: xarray/backends/rasterio_.py (hunk 15 above)
Comment (jhamman) [inline xarray/backends/rasterio_.py]: @NicWayand -

I think you can basically drop this in replacing the current `RasterioArrayWrapper`:

```Python
class RasterioArrayWrapper(NdimSizeLenMixin):
    """Mixin class RasterIO datasets for making wrappers of N-dimensional 
    arrays that conform to the ndarray interface required for the data
    argument to Variable objects.
    A subclass should set the `array` property and override one or more of
    `dtype`, `shape` and `__getitem__`.
    """
    def __init__(self, ds):
        self.ds = ds
    
    @property
    def dtype(self):
        if len(set(self.ds.dtypes[0])) != 1:
            raise ValueError(
                'Can only handle Rastio dataset with all bands having the same type')
        return np.dtype(self.ds.dtypes[0])

    @property
    def shape(self):
        return self.ds.shape

    def __array__(self, dtype=None):
        '''Not sure if this will work as is'''
        return np.asarray(self[...], dtype=dtype)

    def __getitem__(self, key):
        band = range(self.shape[0])[key[0]]
        window = []
        for win in key[1:]:
            if instance(win, slice):
                window.append((win.start, win.stop))
            elif isinstance(win, int):
                window.append((win, win + 1))
            else:  # integer ndarray
                window.append((win.min(), win.max()))
        raw_data = self.ds.read(band, window=window)
        # now, fix up raw_data to conform to numpy indexing conventions
        # - drop axes for integer band/windows
        # - stride if windows are slices with win.step != 1
        # - subset if window is an integer ndarray

    def __repr__(self):
        return '%s(array=%r)' % (type(self).__name__, self.ds)
```

I'm not sure the `__array__` will work as it is until we get the `__get_item__` fully functional.

Comment (shoyer) [inline xarray/backends/rasterio_.py]: I would suggest getting something working here, with an eye toward keeping it self-contained. Later, we can try to port it upstream to rasterio.

**Comment by fmaussion** [comment]
Hi, unless there are objections I'd like to try to finish this PR. I just forked @NicWayand 's branch and started to look at it, and I hope to get something done during next week or so...

**Comment by NicWayand** [comment]
Hi @fmaussion, no objections here. I got it working just barely for my project, and won't have time in the near future to devote to wrap this up.

**Comment by byersiiasa** [comment]
@gidden you might be interested in this

**Comment by gidden** [comment]
@byersiiasa indeed I am, in fact I'm reviewing #1260 which overtook this one =)

**Comment by fmaussion** [comment]
Closing this in favor of https://github.com/pydata/xarray/pull/1260
